src/protocol/sender.rs

```diff
@@ -0,0 +1,622 @@
+            },
+            seq_number: self.get_new_sequence_number(),
+            timestamp: self.time_base.timestamp_from(time),
+            payload,
+        };
+
+        self.buffer.push_back(packet)
+    }
+
+    pub fn pop_front(&mut self) -> Option<DataPacket> {
+        self.buffer.pop_front()
+    }
+
+    pub fn front(&self) -> Option<&DataPacket> {
+        self.buffer.front()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.buffer.is_empty()
+    }
+
+    pub fn latest_message_number(&self) -> MsgNumber {
+        self.next_message_number - 1
+    }
+
+    pub fn latest_seqence_number(&self) -> SeqNumber {
+        self.next_seq_number - 1
+    }
+
+    pub fn next_seqence_number(&self) -> SeqNumber {
+        self.next_seq_number
+    }
+
+    pub fn timestamp_now(&self) -> TimeStamp {
+        self.time_base.timestamp_now()
+    }
+
+    /// Gets the next available message number
+    fn get_new_message_number(&mut self) -> MsgNumber {
+        self.next_message_number += 1;
+        self.next_message_number - 1
+    }
+
+    /// Gets the next avilabe packet sequence number
+    fn get_new_sequence_number(&mut self) -> SeqNumber {
+        // this does looping for us
+        self.next_seq_number += 1;
+        self.next_seq_number - 1
+    }
+}
+
+struct SendBuffer {
+    /// The buffer to store packets for retransmision, sorted chronologically
+    buffer: VecDeque<DataPacket>,
+
+    /// The first sequence number in buffer, so seq number i would be found at
+    /// buffer[i - first_seq]
+    first_seq: SeqNumber,
+}
+
+impl SendBuffer {
+    pub fn new(settings: &ConnectionSettings) -> Self {
+        Self {
+            buffer: Default::default(),
+            first_seq: settings.init_seq_num,
+        }
+    }
+
+    pub fn release_acknowledged_packets(&mut self, acknowledged: SeqNumber) {
+        while acknowledged > self.first_seq {
+            self.buffer.pop_front();
+            self.first_seq += 1;
+        }
+    }
+
+    pub fn get<'a, I: Iterator<Item = SeqNumber> + 'a>(
+        &'a self,
+        numbers: I,
+    ) -> impl Iterator<Item = Result<&'a DataPacket, SeqNumber>> + 'a {
+        numbers.map(
+            move |number| match self.buffer.get((number - self.first_seq) as usize) {
+                Some(p) => Ok(p),
+                None => Err(number),
+            },
+        )
+    }
+
+    pub fn push_back(&mut self, data: DataPacket) {
+        self.buffer.push_back(data);
+    }
+}
+
+pub struct LossList {
+    pub list: VecDeque<DataPacket>,
+}
+
+impl LossList {
+    pub fn new(_settings: &ConnectionSettings) -> Self {
+        Self {
+            list: VecDeque::new(),
+        }
+    }
+
+    pub fn push_back(&mut self, packet: DataPacket) {
+        self.list.push_back(packet);
+    }
+
+    pub fn pop_front(&mut self) -> Option<DataPacket> {
+        self.list.pop_front()
+    }
+
+    pub fn remoeve_acknowledged_packets(&mut self, acknowledged: SeqNumber) -> u32 {
```

Comment:
I think this is a typo

Also a comment saying what this returns would be good, it's not obvious from the signature.